tensorwaves/physics/helicityformalism/amplitude.py

```diff
@@ -170,238 +223,292 @@ def _initialize_parameters(self, recipe: dict) -> None:
                 par["Value"], name=par["Name"], dtype=tf.float64
             )
 
-    def _get_normalization_data(self) -> Optional[dict]:
+    def get_normalization_data(self) -> Tuple[dict, float]:
+        """Return phase space dataset and its volume."""
         if self._phsp_data.size == 0:
             raise Exception(
                 "No phase space sample given! This is required for the "
                 "normalization."
             )
-        return self._phsp_data
+        return (
+            self._kinematics.convert(self._phsp_data),
+            self._kinematics.phase_space_volume,
+        )
 
 
 class _NormalizedIntensity:
-    def __init__(self, recipe: dict, builder, kinematics):
-        self._model_data = builder._create_intensity(recipe["Intensity"])
-        self._model_phsp = builder._create_intensity(recipe["Intensity"])
-        dataset = kinematics.convert(builder._get_normalization_data())
-        # its important to convert the dataset to tf tensors (memory footprint)
-        self._norm_dataset = {x: tf.constant(y) for x, y in dataset.items()}
-        self._norm_volume = atfi.const(kinematics.phase_space_volume)
+    def __init__(
+        self,
+        unnormalized_intensity: Callable,
+        norm_dataset: dict,
+        norm_volume: float = 1.0,
+    ) -> None:
+        self._model = unnormalized_intensity
+        self._norm_dataset = norm_dataset
+        self._norm_volume = norm_volume
 
     @tf.function
-    def __call__(self, dataset):
+    def __call__(self, dataset: dict) -> tf.Tensor:
         normalization = tf.multiply(
-            self._norm_volume,
-            tf.reduce_mean(self._model_phsp(self._norm_dataset)),
+            self._norm_volume, tf.reduce_mean(self._model(self._norm_dataset)),
         )
-        return tf.divide(self._model_data(dataset), normalization)
+        return tf.divide(self._model(dataset), normalization)
+
+
+def _create_normalized_intensity(
```

Comment:
I really don't see how something like this can be avoided or what is so bad about this. Do you have a small pseudocode example on how you imagine this?

Reply:
Let me think about this longer. It's just that this "`_SomeClass` followed by `_create_some_class`" seems rather repetitive, so that the only way to expand this code is copy pasting. But I admit that what these create functions and classes do internally, is quite different (apart from `_create_incoherent_intensity` and `_create_coherent_intensity`).

The nested structure of the recipe imho also just begs for some recursive construction through polymorphism: each of the `_SomeClass`es here is a node in the intensity recipe that can contain other nodes or is an endpoint (here: `_HelicityDecay`). Or maybe a constructor that takes a recipe `dict` itself as argument (so that you don't need that `_create_some_class`)...? But for the moment I also don't see how to do this more nicely.

Of course, the eventual functionality is the same, but in the current code it just takes time for instance to untangle and understand why for instance an `IntensityBuilder` needs to be passed on to the constructor of each `_SomeClass` while their `_create_some_class` factory methods are in turn used in the constructor of that very same `IntensityBuilder`.

So as said, I approved the PR, it's just something I want to keep in mind and think about.

Reply:
I believe what you are looking for is what it was like before (maybe with a few adjustments). So there the creation of for example `_IncoherentIntensity` was done via the recipe in the constructor.
As I tried to make it clear, the reason I changed this is, because I wanted to have an easy way to test the underlying computation code. In the old design this was only possible via recipes, which are not very convenient. Now the `_IncoherentIntensity` callable is directly created from its computation building blocks (list of functions that will be added). That way its quite easy to test the correctness of the computations. See the test file.